### PR TITLE
Add default logging support

### LINKED ISSUE #608: Add logging support to Pyro
Using the `print` function is not ideal for multiple reasons:
 - It does not give us enough control and granularity over logging. A mature model will need less logging, whereas we need more of it while debugging or testing a model. Python's logging library provides us this functionality, as we can configure the logging level.
 - The users of Pyro may want to configure their own handler (e.g. email the user if the model throws an error), and we currently do not have a way of doing that.

Right now, we are mostly using the `print` method in the tests. However, since MCMC algorithms will be outputting verbose diagnostic messages or warnings, we should have a logger handle that rather than printing it.

### PR DESCRIPTION
**Pull request by neerajprad**
Fixes #608. Refer to the task for more details. This is needed so that MCMC algorithms can output diagnostic messages to the logger.

### CODE REVIEW

File: tests/__init__.py [hunk 1]
```diff
@@ -1 +1,14 @@
 from __future__ import absolute_import, division, print_function
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+# create log handler for tests
+handler = logging.StreamHandler()
+formatter = logging.Formatter('%(levelname)s - %(message)s')
```
Comment (fritzo) [inline tests/__init__.py]: I'm fond of `"%(levelname).1s %(name)s %(message)s"` if it's all the same to you.

File: tests/poutine/test_mapdata.py [hunk 2]
```diff
@@ -293,39 +285,45 @@ def irange_cuda_model(subsample_size):
...
     assert tr.batch_log_pdf().is_cuda
 
 
-@pytest.mark.parametrize('model', [
-    iarange_model,
-    irange_model,
-    map_data_vector_model,
-    map_data_iter_model,
-], ids=['iarange', 'irange', 'map_data_vector', 'map_data_iter'])
+@pytest.mark.parametrize(
```
Comment (fritzo) [inline tests/poutine/test_mapdata.py]: Please either revert the autoformat changes or split them into a separate PR.

Comment (neerajprad) [reply tests/poutine/test_mapdata.py]: My bad, this should have been two different commits. Will revert and push again.

File: tests/__init__.py (hunk 1 above)
Comment (neerajprad) [reply tests/__init__.py]: Done.

Comment (neerajprad) [reply tests/__init__.py]: Btw..this is for the root logger for the test. The test output will look as follows (I added a newline before message). Seems a little cluttered to me. What do you think?


```
I tests.nn.test_clipped_nn 
 softmax_ps = Variable containing:
 0.2689  0.7311
[torch.FloatTensor of size 1x2]

.I tests.nn.test_clipped_nn 
 epsilon = 1e-05
I tests.nn.test_clipped_nn 
 softmax_ps = Variable containing:
 0.2689  0.7311
[torch.DoubleTensor of size 1x2]
```

Comment (fritzo) [reply tests/__init__.py]: It would be cleaner to log `softmax_ps.data.cpu().numpy()`, but that seems to be an issue for the user rather than the log handler.

Comment (neerajprad) [reply tests/__init__.py]: That was just a random sample that I picked. Changing this to a tab instead of a newline.

**Comment by eb8680** [comment]
Nice!  Also a prerequisite to integrating ideas from #20.

**Comment by neerajprad** [comment]
> this is great and definitely needed, esp as the system gets more complex.
all the assert messages in the code probably should be changed to caught exceptions with a log.error message

We should still raise an exception if we have entered into a completely unexpected state so that we terminate early and throw a meaningful exception, but we can definitely supplement it with logging to output other information that may be useful for debugging.

File: tests/infer/test_gradient.py [hunk 3]
```diff
@@ -1,5 +1,7 @@
 from __future__ import absolute_import, division, print_function
 
+import logging
```
Comment (fritzo) [inline tests/infer/test_gradient.py]: I'm used to run `pytest -s` to see this test output. How can I invoke `pytest` after this PR to see the logging output? Is it something like 'PYRO_LOG_LEVEL=1 pytest`?

Comment (neerajprad) [reply tests/infer/test_gradient.py]: The test root logger has the default level set to `DEBUG` (we can change this later depending on some environment flag) so you shouldn't need to do anything else. `pytest -s` should work fine.

**Comment by jpchen** [comment]
> We should still raise an exception if we have entered into a completely unexpected state 
Yes sorry I meant replace/add the message not the assert.  Basically exceptions should be greppable in the logs since when you run things on remote machines a lot of times you don't have easy access to the output history

**Comment by neerajprad** [comment]
This is good to merge now.
